addons/godot-next/references/inflector.gd

```diff
@@ -0,0 +1,269 @@
+#     vocabulary = Inflector.Vocabulary.build_default_vocabulary()
+#     Note:
+#     - Empty vocabulary can be manually configured with custom rules.
+#     - Default vocabulary will apply convertions for english language.
+# - Get Vocabulary:
+#     vocabulary = inflector.get_vocabulary() # reference to the vocabulary used by the inflector.
+# - Add Rules:
+#     vocabulary.add_plural(rule, replacement) # adds convertion rule to plural
+#     vocabulary.add_singular(rule, replacement) # adds convertion rule to singular
+#     vocabulary.add_irregular(rule, replacement) # adds irregular convertion
+#     vocabulary.add_uncountable(word) # add unconvertable word
+#     Note:
+#     - 'rule' is a String with regex syntax.
+#     - 'replacement' is a String with regex systax.
+# - Using Inflector:
+#     inflector.pluralize(word, singular = true) # returns the plural of the word
+#     inflector.singularize(word, plural = true) # returns the singular of the word
+#     Note:
+#     - The extra argument is used if there is no ensurance the given word is a singular or plural.
+extends Reference
+class_name Inflector
+
+##### CLASSES #####
+
+class Rule:
+	
+	##### PROPERTIES #####
+	
+	var _regex: RegEx
+	var _replacement: String
+	
+	##### NOTIFICATIONS #####
+	
+	func _init(p_rule: String, p_replacement: String):
+		_regex = RegEx.new()
+		#warning-ignore:return_value_discarded
+		_regex.compile(p_rule)
+		_replacement = p_replacement
+	
+	##### PUBLIC METHODS #####
+	
+	func apply(p_word: String):
+		if not _regex.search(p_word):
+			return null
+		return _regex.sub(p_word, _replacement)
+
+class Vocabulary:
+	
+	##### PROPERTIES #####
+	
+	var _plurals := [] setget, get_plurals
+	var _singulars := [] setget, get_singulars
+	var _uncountables := [] setget, get_uncountables
+	
+	##### PUBLIC METHODS #####
+	
+	func get_plurals():
+		return _plurals
+	
+	func get_singulars():
+		return _singulars
+	
+	func get_uncountables():
+		return _uncountables
+	
+	func add_plural(p_rule: String, p_replacement: String) -> void:
+		_plurals.append(Rule.new(p_rule, p_replacement))
+	
+	func add_singular(p_rule: String, p_replacement: String) -> void:
+		_singulars.append(Rule.new(p_rule, p_replacement))
+	
+	func add_irregular(p_singular: String, p_plural: String, p_match_ending: bool = true) -> void:
+		if p_match_ending:
+			var sfirst = p_singular[0]
+			var pfirst = p_plural[0]
+			add_plural("(" + sfirst + ")" + p_singular.lstrip(sfirst) + "$", "$1" + p_plural.lstrip(pfirst))
+			add_singular("(" + pfirst + ")" + p_plural.lstrip(pfirst) + "$", "$1" + p_singular.lstrip(sfirst))
+		else:
+			add_plural("^%s$" % p_singular, p_plural)
+			add_singular("^%s$" % p_plural, p_singular)
+	
+	func add_uncountable(p_word: String) -> void:
+		_uncountables.append(p_word.to_lower())
+	
+	func is_uncountable(p_word: String) -> bool:
+		return _uncountables.has(p_word.to_lower())
+	
+	static func build_default_vocabulary() -> Vocabulary:
```

Comment:
The vast majority of this function is a bunch of the same method call getting executed over and over again. I would rather see you write a single method call takes in an array (or an array of arrays or something, for each sequence of parameters) and then just execute a for loop within the method that handles everything. Don't necessarily get rid of the methods you have there now (or change them) as we would still want people to be able to edit, on a single Rule basis, the current Vocabulary. But, for this function, you can increase the performance considerably just by changing the way you are "iterating" over the data, minimizing the number of scripting API callbacks you trigger.

Reply:
I hate it when I can't follow your ideas... sorry you'll have to explain it another way :/

Reply:
Okay, so you have something like this...

    func high_level_func():
        low_level_func(data1.new(data11, data12))
        low_level_func(data2.new(data21, data22))
        low_level_func(data3.new(data31, data32))

...so this results in 1 + N (where N = 3) scripting API callbacks that have to jump from the script code to the engine code and back again. You could simplify the logic...

    func high_level_func():
        for a_data in [data1.new(...), data2.new(...), data3.new(...)]:
            low_level_func(a_data)

...and that would improve the readability a lot, but you'd *still* be making all of these low level callback jumps. So, to prevent that, we need to batch up the operations:

    func high_level_func():
        low_level_func([data1.new(...), data2.new(...), data3.new(...)])

at which point you then handle the logic as necessary in the low level function. You also have a few low level funcs, so this will be more readable too

    func build_default_vocabulary():
        low_level_func1([
            <stuff1>,
            <stuff2>, # etc.
        ])

        low_level_func2(...) # etc.

With only needing 3 or 4 low-level batch functions in total. Then you can run your for loops in the lower level methods which converts our 1 + N callbacks to 1 + 1 instead.